httpbakery/agent/agent.go

```diff
@@ -106,121 +123,111 @@ func LoginCookie(req *http.Request) (username string, key *bakery.PublicKey, err
+		username:  username,
+		key:       key,
+	})
 }
 
-// VisitWebPage creates a function that can be used with
-// httpbakery.Client.VisitWebPage. The function uses c to access the
-// visit URL. If no agent-login cookie has been configured for u an error
-// with the cause of ErrNoAgentLoginCookie will be returned. If the login
-// fails the returned error will be of type *httpbakery.Error. If the
-// response from the visitURL cannot be interpreted the error will be of
-// type *UnexpectedResponseError.
-//
-// If using SetUpAuth, it should not be necessary to use
-// this function.
-func VisitWebPage(c *httpbakery.Client) func(u *url.URL) error {
-	return func(u *url.URL) error {
-		err := ErrNoAgentLoginCookie
-		for _, c := range c.Jar.Cookies(u) {
-			if c.Name == cookieName {
-				err = nil
-				break
-			}
+func insertAgent(agents []agent, ag agent) []agent {
+	for i, a := range agents {
+		switch pathCmp(ag.pathParts, a.pathParts) {
+		case -1:
+			agents = append(agents, agent{})
+			copy(agents[i+1:], agents[i:])
+			agents[i] = ag
+			return agents
+		case 0:
+			agents[i] = ag
+			return agents
 		}
-		if err != nil {
-			return errgo.WithCausef(err, http.ErrNoCookie, "cannot perform agent login")
+	}
+	return append(agents, ag)
+}
+
+// pathParts splits a path up into an array of parts.
+func pathParts(path string) []string {
+	return strings.FieldsFunc(path, func(c rune) bool {
+		return c == '/'
+	})
+}
+
+// pathCmp compares two slices of path parts and returns -1 if p1 comes
+// first logically, 1 if p1 comes second logically or 0 if they are
+// equal. The logical ordering used by this function compares each part
+// lexically, if these are equal up to the length of the shortest slice
+// then the longer slice comes first.
+func pathCmp(p1, p2 []string) int {
+	for len(p1) > 0 && len(p2) > 0 {
+		if cmp := strings.Compare(p1[0], p2[0]); cmp != 0 {
+			return cmp
 		}
-		req, err := http.NewRequest("GET", u.String(), nil)
-		if err != nil {
-			return errgo.Notef(err, "cannot create request")
+		p1, p2 = p1[1:], p2[1:]
+	}
+	if len(p1) > 0 {
+		return -1
+	}
+	if len(p2) > 0 {
+		return 1
+	}
+	return 0
+}
+
+// VisitWebPage implements httpbakery.Visitor.VisitWebPage
+func (v *Visitor) VisitWebPage(client *httpbakery.Client, m map[string]*url.URL) error {
+	url := m[httpbakery.UserInteractionMethod]
+	pathParts := pathParts(url.Path)
+	for _, agent := range v.agents[url.Host] {
+		if len(pathParts) < len(agent.pathParts) {
+			continue
 		}
-		resp, err := c.Do(req)
-		if err != nil {
-			return errgo.Notef(err, "cannot perform request")
+		if pathCmp(agent.pathParts, pathParts[:len(agent.pathParts)]) != 0 {
+			continue
 		}
-		defer resp.Body.Close()
-		b, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			logger.Errorf("cannot read response body: %s", err)
-			b = []byte{}
+		logger.Debugf("attempting login to %v using agent %s", url, agent.username)
+		c := &httprequest.Client{
+			Doer: &httpbakery.Client{
```

Comment:
I think this should probably make a shallow copy of client
and just set the key.